verilog/analysis/checkers/numeric_format_string_style_rule.cc

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/numeric_format_string_style_rule.h"
+
+#include <set>
+#include <string>
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "common/analysis/citation.h"
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/token_stream_lint_rule.h"
+#include "common/strings/naming_utils.h"
+#include "common/text/token_info.h"
+#include "verilog/analysis/descriptions.h"
+#include "verilog/analysis/lint_rule_registry.h"
+#include "verilog/parser/verilog_lexer.h"
+#include "verilog/parser/verilog_token_classifications.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+namespace analysis {
+
+using verible::GetStyleGuideCitation;
+using verible::LintRuleStatus;
+using verible::LintViolation;
+using verible::TokenInfo;
+using verible::TokenStreamLintRule;
+
+// Register the lint rule
+VERILOG_REGISTER_LINT_RULE(NumericFormatStringStyleRule);
+
+absl::string_view NumericFormatStringStyleRule::Name() { return "numeric-format-string-style"; }
+const char NumericFormatStringStyleRule::kTopic[] = "number-formatting";
+const char NumericFormatStringStyleRule::kMessage[] =
+  "Formatting string must contain proper style-compilant numeric specifiers.";
+
+std::string NumericFormatStringStyleRule::GetDescription(
+    DescriptionType description_type) {
+  return absl::StrCat("Checks that string literals with numeric format specifiers "
+    "have proper prefixes for hex and bin values and no prefixes for decimal values.");
+}
+
+void NumericFormatStringStyleRule::HandleToken(const TokenInfo& token) {
+  const auto token_enum = static_cast<verilog_tokentype>(token.token_enum());
+  const absl::string_view text(token.text());
+
+  // TODO(fangism): Move it to some utils? It's copy-paste from macro_name_style_rule.cc
+  //     And maybe we could use ShouldRecursivelyAnalyzeToken().
```

Comment:
`ShouldRecursivelyAnalyzeToken` does little more than wrap around `IsLexed`, and just exists to conform to a `std::function` interface where it is used, it may as well just be replaced with an anonymous lambda.  And it's kind of self-documenting.

If I were to refactor this pattern out, it would be something like:

```c++
  if (IsUnlexed(verilog_tokentype(token.token_enum()))) {
    RecursivelyLexTokens([this](const TokenInfo& subtoken){ HandleToken(subtoken); });
    return;
  }
``` 

where

```c++
  // Lexes a token into subtokens while applying a generic function 'func'.
  void RecursivelyLexTokens(const std::function<void(const TokenInfo&)>& func) {
    VerilogLexer lexer(text);
    while (true) {
      const TokenInfo& subtoken(lexer.DoNextToken());
      if (subtoken.isEOF()) break;
      func(subtoken);
    }
  }
```

could be defined in verilog/parser/lexer.cc as a public free-standing function.
How does this look to you?

Reply:
I've updated a comment. Leaving refactoring for you.